README.md
# Jetson Engine
Jetson Engine is a client-server based chess engine framework designed for chess software like ChessBase or Fritz to remotely access UCI-compliant chess engines that run on external devices or computer nodes. This README shows how to compile backend server agent and frontend client UCI engines. More information about Jetson Engine can be found at http://www.ezchess.org/.

Backend server agent supports:
* [Nvidia Jetson Xavier Device](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-xavier-nx/) (arm64, Linux)
* Windows Server (x64, Windows 10)
* Linux Server (x64, Ubuntu Linux)

Frontend client UCI engines supports:
* Windows Host (x64, Windows 10)
* Linux Host (x64, Ubuntu Linux)
* Mac Host (x64, maxOS)

## 1. Downloading Source
```
git clone https://github.com/ezchess/JetsonEngine.git
```

## 2. Building Jetson Backend Agent
Jetson backend servers support three platforms: Nvidia Xavier devices, Windows 10, and Ubuntu Linux.

### 2.1 Xavier and Linux
```
g++ -o jetson_agent agent.cc -lpthread
```

### 2.2 Windows

1. Download and setup [MSYS2 Software Distribution and Building Platform for Windows](https://www.msys2.org/).

2. Under MSYS2, install development tools for MSYS2:
```
unzip -v &> /dev/null || pacman -S --noconfirm unzip
make -v &> /dev/null || pacman -S --noconfirm make
g++ -v &> /dev/null || pacman -S --noconfirm mingw-w64-x86_64-gcc
```

3. Under MSYS2, compile jetson agent:
```
g++ -o jetson_agent.exe agent.cc -liphlpapi -lws2_32 -lpthread -static
```

## 3. Building Jetson Frontend Client
Jetson frontend clients support three platforms: Windows 10, Ubuntu Linux, and Mac OS.

### 3.1 Linux and Mac OS
```
g++ -o jetson_scan client.cc -lpthread
```

### 3.2 Windows
MSYS2 is required for compiling client. Please follow the steps mentioned in section 2.2. After MSYS2 is set up run the following command:
```
g++ -o jetson_scan.exe client.cc -liphlpapi -lws2_32 -lpthread -static
```

## 4. Running Jetson Engine
Please follow the [Jetson Engine User Guide](http://www.ezchess.org/jetson_v2/UserGuide.html) to set up and launch agent and client. For Nvidia Xavier device backend, please follow this [special procedure](http://www.ezchess.org/jetson_v2/XavierUserGuide.html). 

## 5. License
Jetson Engine is a free software. You can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

Jetson Engine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with Jetson Engine. If not, see <http://www.gnu.org/licenses/>.

If you modify this Program, or any covered work, by linking or combining it with third-party libraries, proprietary libraries, or non-GPL compatible libraries, you must obtain licenses from the authors of the libraries before conveying the result work. You are not authorized to redistribute these libraries, whether in binary forms or source codes, individually or together with this Program as a whole unless the terms of the respective license agreements grant you additional permissions.

Author: Evelyn J. Zhu <info@ezchess.org>


CHANGELOG.md
### August 17, 2022: Lc0 benchmark for Jetson AGX Orin
* [JetPack 5.0.2](https://developer.nvidia.com/embedded/downloads) released on 8/15/2022. I benchmark latest [Lc0 development version](https://github.com/LeelaChessZero/lc0) with weights 256x20-t40-1541.pb.gz, and performance reaches 19359 nps. AGX Orin shows 8x performance boost than Xavier NX. [Lc0 networks](https://lczero.org/play/networks/bestnets/) has removed t40-1541. Please use the latest networks for engine evaluations.

```
jetson@agxorin:~/lc0/build/release$ ./lc0 benchmark --backend=cuda-fp16
       _
|   _ | |
|_ |_ |_| v0.30.0-dev+git.dirty built Aug 17 2022
Loading weights file from: ./256x20-t40-1541.pb.gz
Creating backend [cuda-fp16]...
CUDA Runtime version: 11.4.0
Latest version of CUDA supported by the driver: 11.4.0
GPU: Orin
GPU memory: 29.8203 Gb
GPU clock frequency: 1300 MHz
GPU compute capability: 8.7
. . . . . .
. . . . . .

===========================
Total time (ms) : 341324
Nodes searched  : 6607875
Nodes/second    : 19359
```

## Version 2.1

### September 26, 2021: Cloud engine enhancement
* Added customized management port so that port forwarding can re-direct scan/query requests to multiple backend nodes. 

### August 26, 2021: Lc0 benchmark from different JP/cuDNN combinations
* With latest JetPack 4.6 + cuDNN 8.2 released on 8/4/2021, a major performance issue with cuDNN 8.0 is fixed. I benchmark latest [Lc0 0.28.0](https://github.com/LeelaChessZero/lc0/releases/tag/v0.28.0) with weights 256x20-t40-1541.pb.gz. Surprisingly, older JP4.4/cuDNN 7.6 still performs best.
```
JetPack    cuDNN        PowerMode           Performance (Nodes/second)

JP 4.6     cuDNN 8.2    8 (20W, 6-Core)     1693

                        6 (20W, 2-Core)     1936

                        0 (15W, 2-Core)     1777
			
           cuDNN 7.6    6 (20W, 2-Core)     1972
			
JP 4.4     cuDNN 7.6    0 (15W, 2-Core)     2400
```

## Version 2.0

### August 16, 2020: [Jetson Engine v2](http://www.ezchess.org/jetson_v2/UserGuide.html) build 2008.1601
* Fixed a bug in Windows agent.

### August 15, 2020: Jetson Engine v2 build 2008.1505
* Added client support for Linux. UCI engine tested in Ubuntu 18.04.
* Added client support Mac OS. UCI engine tested with HIARCS Chess Explorer in macOS Catalina.
* Fixed a few bugs.

### June 28, 2020: Jetson Engine v2 build 2006.002
* Framework changed to client-server based.
* Backend platform supports Nvidia Jetson Xavier, Windows, and Linux.
* Backend agent module is separated from engine as a standalone program (jetson_agent) to simultaneously communicate with multiple engines.
* Any UCI-compliant engines can be launched in backend server. 
* Frontend client renamed as jetson_scan.exe. UCI engine tested with ChessBase in Windows 10.

## Version 1.0

### June 7, 2020: Add support for [Stockfish](https://stockfishchess.org/)
* [Stockfish source code](https://github.com/official-stockfish/Stockfish) is modified to add backend agent module.
* Frontend client for Stockfish jetsonsf.exe is UCI compliant and tested with ChessBase in Windows 10.

### May 31, 2020: Jetson Engine v1 (aka [Jetson Lc0](http://www.ezchess.org/jetson_v1/JetsonLc0UserGuide.html)) initial release. 
* Backend platform supports [Nvidia Jetson Xavier](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems/jetson-xavier-nx/).
* Backend engine supports [Leela Chess Zero](https://lczero.org/).
* [LCZero source code](https://github.com/LeelaChessZero/lc0) is modified to add backend agent module.
* Frontend client jetsonlc0.exe is UCI compliant and tested with ChessBase in Windows 10.


common/common.h
/*****************************************************************************
 * This file is part of Jetson Engine.
 * 
 * Copyright (C) 2020 Evelyn Zhu
 * 
 * Jetson Engine is a client-server based chess engine framework designed for
 * chess software like ChessBase or Fritz to remotely access UCI-compliant 
 * chess engines that run on external devices or computer nodes.
 *
 * More information about Jetson Engine can be found at http://ezchess.org/
 *
 * Jetson Engine is a free software. You can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jetson Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jetson Engine. If not, see <http://www.gnu.org/licenses/>.
 *
 * If you modify this Program, or any covered work, by linking or
 * combining it with third-party libraries, proprietary libraries, or
 * non-GPL compatible libraries, you must obtain licenses from the 
 * authors of the libraries before conveying the result work. You are
 * not authorized to redistribute these libraries, whether in binary
 * forms or source codes, individually or together with this Program
 * as a whole unless the terms of the respective license agreements
 * grant you additional permissions.
 *
 * The above copyright notice, permission notice, and license agreement
 * information shall be included in all copies or substantial portions
 * of this Program.
 *
 * Author: Evelyn J. Zhu <info@ezchess.org>
 *
 ****************************************************************************/

#ifndef _JET_COMMON_H
#define _JET_COMMON_H

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <pthread.h>
#include <csignal>
#include <time.h>
#include <cstdarg>

#if defined(_WIN32)
	#ifndef _WIN32_WINNT
	#define _WIN32_WINNT 0x0600
	#endif
	#include <winsock2.h>
	#include <ws2tcpip.h>
	#pragma comment(lib, "ws2_32.lib")
	#include <windows.h>
	#include <direct.h>
	#include <io.h>
#else
	#include <sys/socket.h>
	#include <netinet/in.h>
	#include <arpa/inet.h>
	#include <netdb.h>
	#include <unistd.h>
	#include <errno.h>
	#include <fcntl.h> 
	#include <sys/utsname.h>
	#include <sys/stat.h>
	#include <thread>
	#include <cstring>
#endif

#if defined(_WIN32)
	#define IsSockValid(s) ((s) != INVALID_SOCKET)
	#define CloseSocket(s) closesocket(s)
	#define GetSockErrno() (WSAGetLastError())
	#define GetCurrDir _getcwd
	#define SleepMsec(msec) Sleep(msec)
#else
	#define IsSockValid(s) ((s) >= 0)
	#define CloseSocket(s) close(s)
	#define GetSockErrno() (errno)
	#define GetCurrDir getcwd
	#define SleepMsec(msec) this_thread::sleep_for(chrono::milliseconds(msec))
	#define CloseHandle(h) close(h)
	#define FileExists(f) IsFileExist(f)
	#define SOCKET int
	#define HANDLE int
#endif

#define MAX_NAME_LEN				256	//max length for all types of names
#define MAX_NUM_ENGINE				32	//max numbers of engine folders allowed to set in server
#define MAX_NUM_LOGI_PER_ENGINE		64	//max numbers of client connections for each individual

struct EngineEntry;

struct ClientEntry {
	int bIsConnected;
	int bIsDataLogOn;
	char sReqPipe[MAX_NAME_LEN];
	char sRspPipe[MAX_NAME_LEN];
	HANDLE hReqPipe;
	HANDLE hRspPipe;
	SOCKET sock;
	char sIpAddr[MAX_NAME_LEN];
	char sServIpAddr[MAX_NAME_LEN];
	char sEngInstName[MAX_NAME_LEN];//engine instance name
	struct EngineEntry *engine;
	fd_set *pMaster;
} __attribute__((aligned(8)));

struct EngineEntry {
	int bIsAllocated;
	int reserved;
	char sEngineDir[MAX_NAME_LEN];
	char sEngineName[MAX_NAME_LEN];
	char sEngineExeName[MAX_NAME_LEN];
	char sEngienPort[MAX_NAME_LEN];
	char arguments[MAX_NAME_LEN];
	struct ClientEntry clients[MAX_NUM_LOGI_PER_ENGINE];
} __attribute__((aligned(8)));

enum SocketType {
	SOCK_TYPE_MGMT = 1,
	SOCK_TYPE_ENGINE = 2
};

enum OsArch {
	OS_ARCH_UNKNOWN	= 0,
	OS_ARCH_XAVIER_ARM64 = 1,	//Nvidia Xavier Tegra, ARM64
	OS_ARCH_WINDOWS_X64 = 2,	//Windows 10, X86-64
	OS_ARCH_LINUX_X64 = 3		//Ubuntu Linux, X86-64
};

#define RSP_BUFSIZE 8192			//uci command, mgmt command
#define REQ_BUFSIZE	1024			//uci response
#define PIPE_BUFSIZE RSP_BUFSIZE	//pipe read/write
#define QUERY_BUFSIZE 32768			//query response

#define STR_OS_ARCH_WIN		(char *)"Windows X86-64"
#define STR_OS_ARCH_LINUX	(char *)"Linux X86-64"
#define STR_OS_ARCH_XAVIER	(char *)"Xavier ARM64"

#define STR_JRE_HDR_WIN		(char *)"JRE_X64WIN_"
#define STR_JRE_HDR_LINUX	(char *)"JRE_X64LNX_"
#define STR_JRE_HDR_XAVIER	(char *)"JRE_XAVIER_"

#define STR_MGMT_PORT (char *)"53350"

#define BUILD_NUMBER (char *)"v2.2008.1601"

static inline const std::string GetCurrentDateTime()
{
	time_t now = time(0);
	struct tm tstruct;
	char buf[80];
	tstruct = *localtime(&now);
	strftime(buf, sizeof(buf), "%Y-%m-%d.%X", &tstruct);
	return buf;
}

static inline bool IsFileExist(const char *fileName)
{
    std::ifstream inFile(fileName);
    return inFile.good();
}

extern char gsMyHostName[];
extern pthread_mutex_t gLogFileLock;
extern char gsLogFile[];
static inline void JetsonWriteLogs(const char *fmt, ...)
{
	//Please ensure sErrMsg ended with '\n'
	pthread_mutex_lock(&gLogFileLock);
	FILE *p = fopen(gsLogFile, "a");
	if (p!= NULL) {
		std::ostringstream oss;
		oss << GetCurrentDateTime() << " [" << gsMyHostName << "] " << fmt;
		
		va_list args;
		va_start(args, fmt);
		vfprintf(p, oss.str().c_str(), args);
		va_end(args);
		
		fclose(p);
	}
	pthread_mutex_unlock(&gLogFileLock);
}

#endif	//_JET_COMMON_H


uciengines/client.cc
/*****************************************************************************
 * This file is part of Jetson Engine.
 * 
 * Copyright (C) 2020 Evelyn Zhu
 * 
 * Jetson Engine is a client-server based chess engine framework designed for
 * chess software like ChessBase or Fritz to remotely access UCI-compliant 
 * chess engines that run on external devices or computer nodes.
 *
 * More information about Jetson Engine can be found at http://ezchess.org/
 *
 * Jetson Engine is a free software. You can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jetson Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jetson Engine. If not, see <http://www.gnu.org/licenses/>.
 *
 * If you modify this Program, or any covered work, by linking or
 * combining it with third-party libraries, proprietary libraries, or
 * non-GPL compatible libraries, you must obtain licenses from the 
 * authors of the libraries before conveying the result work. You are
 * not authorized to redistribute these libraries, whether in binary
 * forms or source codes, individually or together with this Program
 * as a whole unless the terms of the respective license agreements
 * grant you additional permissions.
 *
 * The above copyright notice, permission notice, and license agreement
 * information shall be included in all copies or substantial portions
 * of this Program.
 *
 * Author: Evelyn J. Zhu <info@ezchess.org>
 *
 ****************************************************************************/

#include "../common/common.h"

using namespace std;

static int gbClientExiting = 0;
static int gbScanNeeded = 0;
static int gbQueryNeeded = 0;

static char gsScanBuffer[RSP_BUFSIZE];
static char gsQueryBuffer[QUERY_BUFSIZE];

static SOCKET gServSock;

#if defined(_WIN32)
static char *gsJetsonScanFile = (char *)"jetson_scan.exe";
#else
static char *gsJetsonScanFile = (char *)"jetson_scan";
#endif

char gsMyHostName[MAX_NAME_LEN] = "UNKNOWN_HOST";
pthread_mutex_t gLogFileLock;
char gsLogFile[MAX_NAME_LEN] = "JetsonErr_";

static void *ClientReciverThread(void *data)
{
	JetsonWriteLogs(">>> Entered receiver thread\n");
	
	if (gbScanNeeded)
		memset(gsScanBuffer, 0, RSP_BUFSIZE);
		
	if (gbQueryNeeded)
		memset(gsQueryBuffer, 0, QUERY_BUFSIZE);

	try {	
		while(1) {
			fd_set reads;
			FD_ZERO(&reads);
			FD_SET(gServSock, &reads);
#if !defined(_WIN32)
			FD_SET(0, &reads);
#endif
			struct timeval timeout;
			timeout.tv_sec = 0;
			timeout.tv_usec = 100000;
				
			if (gbClientExiting)
				break;

			if (select(gServSock+1, &reads, 0, 0, &timeout) < 0) {
				JetsonWriteLogs("select() failed. (%d)\n", GetSockErrno());
				throw runtime_error("select() failed\n");
			}

			if (FD_ISSET(gServSock, &reads)) {
				char sSockReadBuf[RSP_BUFSIZE];
				memset(sSockReadBuf, 0, RSP_BUFSIZE);
				int bytes_received = recv(gServSock, sSockReadBuf, RSP_BUFSIZE, 0);
				if (bytes_received < 1) {
					gbClientExiting = 1;
					break;
				}

				std::string line = sSockReadBuf;
			
				if (gbScanNeeded) {
				   	strncat(gsScanBuffer, sSockReadBuf, RSP_BUFSIZE);

					if (strstr(line.c_str(), "scanisdone"))
				   		break;
				}
				else if (gbQueryNeeded) {
					cout << sSockReadBuf;
					JetsonWriteLogs("%s", sSockReadBuf);
					if (strstr(line.c_str(), "querydone")) {
						gbClientExiting = 1;
						break;
					}
				}
				else
		   			cout << sSockReadBuf;//uci response data to ChessBase
			}
		}//while()

		if (gbScanNeeded) {
			char *sBaseFile = (char *)data;
			char cCurrentPath[FILENAME_MAX];

			if (!GetCurrDir(cCurrentPath, sizeof(cCurrentPath)))
				throw runtime_error("unable to get current path\n");

			cCurrentPath[sizeof(cCurrentPath) - 1] = '\0'; /* not really required */

			char * line = strtok(strdup(gsScanBuffer), "\n");
			while(line) {
				cout << line << endl;
				JetsonWriteLogs("%s\n", line);
   			
				if (strstr(line, "scanisdone"))
					break;
				else {
					ostringstream ossNewFile;
#if defined(_WIN32)
					ossNewFile << cCurrentPath << "\\" << line << ".exe";
#else
					ossNewFile << cCurrentPath << "/" << line;
#endif

					ofstream ifs (ossNewFile.str());
					if (!ifs.is_open()) {
						cout << "ERROR: file " << ossNewFile.str() << " is open" << endl;
						JetsonWriteLogs("ERROR: file %s is open\n", ossNewFile.str().c_str());
					}
					else
					{
						ifs.close();
						ostringstream ossCmdline;

#if defined(_WIN32)						
						ossCmdline << "copy " << cCurrentPath << "\\" << sBaseFile << " " << cCurrentPath << "\\" << line << ".exe";
#else
						ossCmdline << "cp " << cCurrentPath << "/" << sBaseFile << " " << cCurrentPath << "/" << line;
						ossCmdline << "; " << "chmod a+x " << cCurrentPath << "/" << line;
#endif
						JetsonWriteLogs("exec cmd is(%s)\n", ossCmdline.str().c_str());
						cout << "exec cmd is(" << ossCmdline.str() << ")" << endl;
						system(ossCmdline.str().c_str());
						SleepMsec(100);
					}
   				}
				
				line  = strtok(NULL, "\n");
			}
		
			gbClientExiting = 1;
		}
	} catch (exception& e) {		
		JetsonWriteLogs("<<< ERROR: %s", e.what());
	}
	
	JetsonWriteLogs("<<< Exited receiver thread\n");
	
	return NULL;
}

#define STR_JREHDR_SIZE 16
#define STR_OSARCH_SIZE 16
#define STR_IPADDR_SIZE 32
#define STR_TCPPORT_SIZE 16

int main(int argc, char *argv[])
{
	int rc;	
	char sJreHeader[STR_JREHDR_SIZE];
	char sOsArch[STR_OSARCH_SIZE];
	char sServIp[STR_IPADDR_SIZE];
	char sServPort[STR_TCPPORT_SIZE];
	char sEngName[MAX_NAME_LEN];
	char sThisExeFileName[MAX_NAME_LEN];
	
	memset(sServIp, 0, STR_IPADDR_SIZE);
	memset(sServPort, 0, STR_TCPPORT_SIZE);

#if defined(_WIN32)
	WSADATA d;
	if (WSAStartup(MAKEWORD(2, 2), &d)) {
		std::cerr << "Failed to initialize." << endl;
		return 0;
	}
	gethostname(gsMyHostName, sizeof(gsMyHostName));
#else
	struct utsname buffer;
	if (uname(&buffer) != 0) {
		std::cerr << "Error on uname." << endl;
		return 0;
	}
   	strcpy(gsMyHostName, buffer.nodename);
#endif
	//TODO: at this moment we don't pass arguments to the backend as arguments
	//are in fact already configured individually in backend's agent.conf
	//however, we're still debating if it is better to pass down
	
	//get current executable name w/o full path info
	int startPos = 0;
	for (int i=0; i<strlen(argv[0]); i++) {
#if defined(_WIN32)
		if (argv[0][i] == '\\')
#else
		if (argv[0][i] == '/')
#endif
			startPos = i;
	}
	if (startPos > 0)
		startPos++;
	
	strncpy(sThisExeFileName, argv[0]+startPos, MAX_NAME_LEN);
	ostringstream ossLogFile;
	ossLogFile << sThisExeFileName << ".log";
	strncat(gsLogFile, ossLogFile.str().c_str(), MAX_NAME_LEN);
	
	ostringstream oss;
	oss << "Received incoming command line=";
	for (int i=0; i<argc; i++) {
		oss << argv[i] << " ";
	}
	oss << ", ver=" << BUILD_NUMBER;
	JetsonWriteLogs("%s\n", oss.str().c_str());
	
	JetsonWriteLogs("Filename = %s\n", sThisExeFileName);

	/* engine server scan or query */
	if (argc >= 3 &&
		strcmp(sThisExeFileName, gsJetsonScanFile) == 0) {
		string mgmtPortStr = STR_MGMT_PORT;
		if (argc >= 4)
			mgmtPortStr = argv[3];
		if (strcmp(argv[1], "scan") == 0) {
			gbScanNeeded = 1;
			strncpy(sServIp, argv[2], STR_IPADDR_SIZE);
			strncpy(sServPort, mgmtPortStr.c_str(), STR_TCPPORT_SIZE);
			
			printf("scanning server %s on port %s\n", sServIp, sServPort);
		}
		
		if (strcmp(argv[1], "query") == 0) {
			gbQueryNeeded = 1;
			strncpy(sServIp, argv[2], STR_IPADDR_SIZE);
			strncpy(sServPort, mgmtPortStr.c_str(), STR_TCPPORT_SIZE);
			
			printf("query server %s on port %s\n", sServIp, sServPort);
		}
	}
	else if (strcmp(sThisExeFileName, gsJetsonScanFile) == 0) {
		printf("Incorrect syntax\n");
		printf("To scan agent run: jetson_scan scan <agent ip address> [mgmt_port]\n");
		printf("To query agent run: jetson_scan query <agent ip address> [mgmt_port]\n");
		printf("Note: mgmt_port is optional. Default port = 53350.\n");
		printf("Example:\n");
		printf("jetson_scan scan 192.168.55.1\n");
		printf("jetson_scan scan 192.168.55.1 61234\n");
		printf("jetson_scan query 192.168.55.1\n");
		printf("jetson_scan query 192.168.55.1 61234\n");
		return 0;
	}
	else {	//JRE engine
		for (int i=0; i<strlen(sThisExeFileName); i++) {
			if (sThisExeFileName[i] == '_')
				sThisExeFileName[i] = ' ';
		}

		sscanf(sThisExeFileName, "%s %s %s %s %s", sJreHeader, sOsArch, sServIp, sServPort, sEngName);
	}
	
	try {	
		struct addrinfo localAddr;
		memset(&localAddr, 0, sizeof(localAddr));
		localAddr.ai_socktype = SOCK_STREAM;
	
		struct addrinfo *pPeerAddr;
		if (getaddrinfo(sServIp, sServPort, &localAddr, &pPeerAddr)) {
			JetsonWriteLogs("getaddrinfo() failed. (%d)\n", GetSockErrno());
			throw runtime_error("getaddrinfo() failed\n");
		}

		char sAddrBuf[100];
		char sServBuf[100];
		getnameinfo(pPeerAddr->ai_addr, pPeerAddr->ai_addrlen,
			sAddrBuf, sizeof(sAddrBuf),
			sServBuf, sizeof(sServBuf),
			NI_NUMERICHOST);

		gServSock = socket(pPeerAddr->ai_family,
			pPeerAddr->ai_socktype, pPeerAddr->ai_protocol);
		if (!IsSockValid(gServSock)) {
			JetsonWriteLogs("socket() failed. (%d)\n", GetSockErrno());
			throw runtime_error("socket() failed\n");
		}

		if (connect(gServSock,
				pPeerAddr->ai_addr, pPeerAddr->ai_addrlen)) {
			JetsonWriteLogs("connect() failed. (%d)\n", GetSockErrno());
			throw runtime_error("connect() failed\n");
		}
		freeaddrinfo(pPeerAddr);
	
		void *pThreadData = (gbScanNeeded ? (void *)sThisExeFileName : NULL);

		pthread_t clientReceiverThreadId = 0;
		rc = pthread_create(&clientReceiverThreadId, NULL,  ClientReciverThread, pThreadData);
		if (rc != 0) {
			throw std::runtime_error("Unable to create recv_thread\n");
		}
	
		if (gbScanNeeded || gbQueryNeeded) {
			send(gServSock, argv[1], strlen(argv[1]), 0);
		
			while(1) {
				SleepMsec(1000);
				if (gbClientExiting)
					break;
			}
		}
		else {
			std::cout.setf(std::ios::unitbuf);
			std::string line;
			while (std::getline(std::cin, line)) {
				if (gbClientExiting)
					throw std::runtime_error("Connection closed by Jetson device\n");			
		
				int bytesSent = send(gServSock, line.c_str(), line.length(), 0);
			
				SleepMsec(300);

				if(!strncmp(line.c_str(), "quit", 4)) {
					gbClientExiting = 1;
					SleepMsec(500);
					break;
				}
			}
		}

		CloseSocket(gServSock);

#if defined(_WIN32)
		WSACleanup();
#endif

		return 0;
	
	} catch (std::exception& e) {
		JetsonWriteLogs("Unhandled exception: %s", e.what());
		
		CloseSocket(gServSock);
#if defined(_WIN32)
		WSACleanup();
#endif
		std::cerr << "Unhandled exception: " << e.what() << endl;
	}
}
